Switch ON objfind QA

debora-pe opened the pull request:
As titled

This works only for multi-slit reduction.
Here is how the plot looks like
![1Dextraction](https://user-images.githubusercontent.com/16786072/140808534-0b98472e-dadf-4a5b-bb09-412bcccc1168.png)
.

debora-pe committed:
switch on qa

profxj commented on pypeit/reduce.py:
```diff
@@ -1123,6 +1125,8 @@ def find_objects_pypeline(self, image, std_trace=None,
             # TODO we need to add QA paths and QA hooks. QA should be
```
maybe we can remove this TODO?

profxj reviewed (approved):
Looks great, thx!

debora-pe committed:
fix a few bugs

debora-pe committed:
fix another small crash

kbwestfall commented on pypeit/coadd2d.py:
```diff
@@ -470,7 +471,7 @@ def reduce(self, pseudo_dict, show=None, show_peaks=None):
         #  2d coadd reduce
         sobjs_obj, nobj, skymask_init = redux.find_objects(
-            sciImage.image, show_peaks=show_peaks,
+            sciImage.image, show_peaks=True,
```
Does this mean it's no longer possible to toggled off the plots?

kbwestfall commented on pypeit/core/extract.py:
```diff
@@ -1517,8 +1519,12 @@ def objfind(image, thismask, slit_left, slit_righ, inmask=None, fwhm=3.0, use_us
+            plt.show()
+        elif outfile_objprof is not None:
+            plt.savefig(outfile_objprof, dpi=400)
+            plt.close('all')
```
I think the logic here is difficult.  I.e., someone could set `show_peaks=True`, but `debug_all=False` and not provide an `outfile_objprof` and then `pyplot` will have created something that never gets flushed, which will mess up the next plot being made.  It would be good to clean up how the QA is done for this function, but, barring that, I would put the `plt.close('all')` call outside the `if...elif` block.

kbwestfall reviewed (approved):
I would need to dig in more, but it feels like it's now very difficult to know how to toggle the QA.  I guess the point of the PR is to set it as always on, but it leaves hooks that may or may not work or even be relevant.  It may be more effort than it's worth to clean all this up in this PR (or it maybe easier than my brief impression), but we should raise an issue so that we're tracking it.

In any case, I'm approving now so as not to hold up moving forward.

jhennawi commented on pypeit/core/extract.py:
```diff
@@ -1517,8 +1519,12 @@ def objfind(image, thismask, slit_left, slit_righ, inmask=None, fwhm=3.0, use_us
-        viewer, ch = display.show_image(image*(thismask*inmask))
-        plt.show()
+        if debug_all:
+            viewer, ch = display.show_image(image*(thismask*inmask))
```
I'm also not following the logic here. Is the issue whether or not the plotting is interactive? Note that you can save the file to a figure, and also show it. In other words, you could simply have

plt.title(qa_title + ': Slit# {:d}'.format(specobj_dict['SLITID']))
if outfile_objprof is not None: 
    plt.savefig(outfile_objprof, dpi=400)
if debug_all: 
   viewer, ch = display.show_image(image*(thismask*inmask))
   plt.show()

plt.close('all)

jhennawi commented on pypeit/core/extract.py:
```diff
@@ -1517,8 +1519,12 @@ def objfind(image, thismask, slit_left, slit_righ, inmask=None, fwhm=3.0, use_us
         plt.title(qa_title + ': Slit# {:d}'.format(specobj_dict['SLITID']))
-        viewer, ch = display.show_image(image*(thismask*inmask))
-        plt.show()
+        if debug_all:
```
So if outfile_objprof is not None, but show_peaks is False, then the plot does not get made. I think we need to think through the QA logic here. 

The point is we either want to do some full interactive plotting, or we want to make QA files and we should have the flexibility to do one, the other, or both.

jhennawi commented on pypeit/pypeit.py:
```diff
@@ -804,7 +804,7 @@ def objfind_one(self, frames, det, bg_frames, std_outfile=None):
 
         # Do it
-        global_sky, sobjs_obj, skymask = self.redux.run_objfind(std_trace=std_trace, show_peaks=self.show)
+        global_sky, sobjs_obj, skymask = self.redux.run_objfind(std_trace=std_trace)
```
Why are you disabling self.show? I think we still want that option, which as per my comment above can be achieved if we specify whether we want interactive QA, or the QA plot to disk.

jhennawi commented on pypeit/reduce.py:
```diff
@@ -371,7 +369,7 @@ def run_objfind(self, std_trace=None, show_peaks=False):
         self.sobjs_obj, self.nobj, skymask_init = \
             self.find_objects(self.sciImg.image, std_trace=std_trace,
-                              show_peaks=show_peaks,
+                              show_peaks=self.par['reduce']['findobj']['skip_second_find'],
```
Let's discuss these options. I think we want to have the option to have interactive plotting of both object finding steps since that is useful for debugging. Currently you are disabling that. This does not exclude the option to write out the result of the second object finding to a QA file, and so we should also have the option to do that.

jhennawi commented on pypeit/scripts/ql_keck_mosfire.py:
```diff
@@ -120,7 +120,7 @@ def run_pair(A_files, B_files, caliBrate, spectrograph, det, parset, show=False,
     #     std_trace=std_trace, return_negative=True, show_peaks=show)
 
-    global_sky, sobjs_obj, skymask = redux.run_objfind(std_trace=std_trace, show_peaks=show)
+    global_sky, sobjs_obj, skymask = redux.run_objfind(std_trace=std_trace)
```
Is there a reason why this has been disabled? Please do not remove the previous show_peaks functionality.

jhennawi reviewed (changes requested):
Hi @deborape, 

Many thanks for getting this going. Please see my comments on the code below. 

Two major comments: 

1) You appear to be changing the previous interactive functionality, which is in my opinion not desirable. The previous interactive functionality to show the interactive object finding plot for the first and second object finding is not at odds with your goal of writing out the result of the second object finding plot to a QA file. I think we just need to think a little more about the parameters and the logic. 

2) I don't understand why this should only work for multislit. Have you tested it on echelle? Given that this is all at your fingertips now, whereas someone trying to implement echelle in the future would have to climb the learning curve you climbed again, I ask that you please try to implement echelle. If that for some unforeseen reason turns out to be complicated, then we can postpone. 

Thanks!
Joe

debora-pe committed:
Merge branch 'develop' into objfind_qa

debora-pe committed:
new approach

debora-pe committed:
very minor aesthetic change

debora-pe commented:
@jhennawi , @kbwestfall , @profxj  I have updated the work for this PR. Now it works for both multislit and echelle, so please take a look.

Tests pass.

```
Dev Suite
--- PYPEIT DEVELOPMENT SUITE PASSED 136/136 TESTS (Masters ignored) ---
Testing Started at 2021-12-10T16:58:30.344252
Testing Completed at 2021-12-11T03:36:42.841274
Total Time: 10:38:12.497022
```
Unit tests
```
pytest . -W ignore
=============================== test session starts ================================
platform darwin -- Python 3.8.5, pytest-6.1.2, py-1.9.0, pluggy-0.13.1
rootdir: /Users/dpelliccia/PypeIt/PypeIt, configfile: setup.cfg
collected 304 items                                                                

test_alignments.py .                                                         [  0%]
test_arc.py .                                                                [  0%]
test_archive.py ..                                                           [  1%]
test_arcimage.py ..                                                          [  1%]
test_biasframe.py ..                                                         [  2%]
test_bitmask.py ....                                                         [  3%]
test_bpmimage.py ..                                                          [  4%]
test_bspline.py ...............                                              [  9%]
test_buildimage.py ..                                                        [ 10%]
test_calibrations.py .......                                                 [ 12%]
test_chk_calibs.py ..                                                        [ 13%]
test_coadd.py .                                                              [ 13%]
test_collate_1d.py ............                                              [ 17%]
test_cooked.py .                                                             [ 17%]
test_coremeta.py .                                                           [ 18%]
test_datacontainer.py ........                                               [ 20%]
test_detector.py ...                                                         [ 21%]
test_display.py .                                                            [ 22%]
test_edgetrace.py .                                                          [ 22%]
test_fitting.py ....                                                         [ 23%]
test_flatfield.py .                                                          [ 24%]
test_flexure.py ..                                                           [ 24%]
test_flux.py ....                                                            [ 25%]
test_fluxspec.py .....                                                       [ 27%]
test_frametype.py ..                                                         [ 28%]
test_history.py ....                                                         [ 29%]
test_load.py .                                                               [ 29%]
test_load_images.py ................                                         [ 35%]
test_maskdesign_matching.py ...                                              [ 36%]
test_masterframe.py .                                                        [ 36%]
test_match.py ....                                                           [ 37%]
test_metadata.py ......                                                      [ 39%]
test_moment.py ......                                                        [ 41%]
test_msgs.py ..                                                              [ 42%]
test_opticalmodel.py ....                                                    [ 43%]
test_parse.py ...                                                            [ 44%]
test_pca.py ........                                                         [ 47%]
test_processimage.py ..                                                      [ 48%]
test_procimg.py .....                                                        [ 49%]
test_pydl.py .                                                               [ 50%]
test_pypeit.py ..                                                            [ 50%]
test_pypeitimage.py ..                                                       [ 51%]
test_pypeitpar.py ................................                           [ 61%]
test_pypeitsetup.py ........                                                 [ 64%]
test_qa.py .                                                                 [ 64%]
test_runpypeit.py ..                                                         [ 65%]
test_scienceimage.py ...                                                     [ 66%]
test_scripts.py .................                                            [ 72%]
test_setups.py ......................                                        [ 79%]
test_skysub.py ..                                                            [ 79%]
test_slitmask.py ...                                                         [ 80%]
test_slittrace.py ....                                                       [ 82%]
test_spec2dobj.py ......                                                     [ 84%]
test_specobj.py .......                                                      [ 86%]
test_specobjs.py .....                                                       [ 88%]
test_spectrographs.py ....................                                   [ 94%]
test_syncspec.py .                                                           [ 95%]
test_telluric.py ..                                                          [ 95%]
test_traceimage.py .                                                         [ 96%]
test_utils.py ...                                                            [ 97%]
test_vcorr.py ..                                                             [ 97%]
test_wavemodel.py .                                                          [ 98%]
test_wavetilts.py ....                                                       [ 99%]
test_wvcalib.py ..                                                           [100%]

========================= 304 passed in 1324.24s (0:22:04) =========================
```

debora-pe committed:
very minor aesthetic change

debora-pe commented:
Examples of the QA plots.

![pos_DE 20170425 50487-dra11_DEIMOS_20170425T140121 014_D07_S0091_obj_prof](https://user-images.githubusercontent.com/16786072/145690861-d090dd6a-063c-4713-9f8f-eb952da35e9c.png)
![pos_cN20170331S0221-cN20170331S0217-pisco_D01_S0003_obj_prof](https://user-images.githubusercontent.com/16786072/145690913-d8c64d44-1761-49cd-9e0f-0da06761b917.png)

profxj commented on pypeit/core/extract.py:
```diff
@@ -1263,6 +1264,10 @@ def objfind(image, thismask, slit_left, slit_righ, inmask=None, fwhm=3.0, use_us
                                 'objtype': 'unknown', 'pypeline': 'unknown'}
+        objfindQA_dict: dict, default = None
+            Dictionary containing meta-data for the object finding QA:
+            SAVE_TO_DISK, BASENAME, OUTDIR
```
Can you briefly describe what these 3 keys are for?

profxj commented on pypeit/core/extract.py:
```diff
@@ -1920,6 +1940,9 @@ def ech_objfind(image, ivar, slitmask, slit_left, slit_righ, order_vec, maskslit
             Boxcar radius in arcsec
+        objfindQA_dict: dict, default = None
+            Dictionary containing meta-data for the object finding QA:
+            SAVE_TO_DISK, BASENAME, OUTDIR
```
Add `SLITORD_ID` too?

profxj commented on pypeit/reduce.py:
```diff
@@ -1077,6 +1080,10 @@ def find_objects_pypeline(self, image, std_trace=None,
             Generate QA  showing traces identified. Requires an open ginga RC
             modules window
         show : :obj:`bool`, optional
+        save_objfindQA : :obj:`bool`, optional
```
Should it simply default to `False`?

jhennawi replied:
I think it should default to True.

debora-pe replied:
Now default to True.

profxj commented on pypeit/reduce.py:
```diff
@@ -1337,8 +1346,8 @@ def get_positive_sobj(self, specobjs, iord):
                               show=False, show_peaks=False, show_fits=False,
-                              show_trace=False, debug=False,
-                              manual_extract_dict=None):
+                              show_trace=False, save_objfindQA=None, neg=False,
```
as above, default `save_objfindQA` to `False`?

profxj reviewed (approved):
Just a few more comments from me.
Nothing all that substantial

jhennawi commented on pypeit/core/extract.py:
```diff
@@ -1497,8 +1502,18 @@ def objfind(image, thismask, slit_left, slit_righ, inmask=None, fwhm=3.0, use_us
     else:
         nobj_reg = 0
 
+    # Set objfind QA filename
```
My suggestion is to not pass in this dictionary and synthesise the filename here as you are doing. Instead run the qa.set_qa_filename code in the routine that calls this core objfind method to generate the filename.

debora-pe replied:
Done

jhennawi commented on pypeit/core/extract.py:
```diff
@@ -1813,7 +1833,7 @@ def ech_objfind(image, ivar, slitmask, slit_left, slit_righ, order_vec, maskslit
                 trim_edg=(5,5), cont_fit=True, npoly_cont=1, show_peaks=False, show_fits=False, show_single_fits=False,
                 show_trace=False, show_single_trace=False, debug=False, show_pca=False,
-                debug_all=False, skymask_by_boxcar=False, boxcar_rad=None):
+                debug_all=False, skymask_by_boxcar=False, boxcar_rad=None, objfindQA_dict=None):
```
Get rid of the dict.

debora-pe replied:
Done

jhennawi commented on pypeit/core/extract.py:
```diff
@@ -1143,7 +1144,7 @@ def objfind(image, thismask, slit_left, slit_righ, inmask=None, fwhm=3.0, use_us
             skymask_nthresh=1.0, specobj_dict=None, cont_fit=True, npoly_cont=1, find_min_max=None,
             show_peaks=False, show_fits=False, show_trace=False, show_cont=False, debug_all=False,
-            qa_title='objfind'):
+            qa_title='objfind', objfindQA_dict=None):
```
See my comment below. You should only be passing in the QA outfile here, not this dictionary.

debora-pe replied:
Done

jhennawi commented on pypeit/core/extract.py:
```diff
@@ -1920,6 +1940,9 @@ def ech_objfind(image, ivar, slitmask, slit_left, slit_righ, order_vec, maskslit
             If True, use the boxcar radius in the skymask
         boxcar_rad: float, optional
             Boxcar radius in arcsec
+        objfindQA_dict: dict, default = None
```
Here it may be necessary to pass in the dict. I'd prefer however that this code just take a string which acts as the prefix for the QA and then you append on the slitord_id to generate that filename in the loop over orders.

debora-pe replied:
Done. It's a string and the `slitord_id` is appended in the loop over the orders.

jhennawi commented on pypeit/reduce.py:
```diff
@@ -1122,9 +1129,10 @@ def find_objects_pypeline(self, image, std_trace=None,
+            objfindQA_dict = {'SAVE_TO_DISK': save_objfindQA, 'BASENAME': self.basename, 'SLITORD_ID': slit_spat,
+                              'OUTDIR': os.path.join(self.par['rdx']['redux_path'], self.par['rdx']['qadir']),
+                              'NEG': neg}
 
```
Run the code to synthesise the filename here ,and then only pass the core method the filename. This has the benefit that you reduce dependencies and failure modes in the core objfind method. For example, a string cannot fail, but a call to qa.set_qa_filename can fail, and that dependency is completely unnecessary since you can generate the filename outside of that routine.

debora-pe replied:
Done.

jhennawi reviewed (changes requested):
Minor changes requested related to simplifying arguments to the core objfind and ech_objfind methods and also reducing dependencies in those methods. 

Can you also post an example of how the QA file hierarchy looks for echelle, and then test the 

run_pypeit -s by hand, since we don't have that in our current development tests.

debora-pe committed:
dealing with conflicts

debora-pe committed:
Merge branch 'develop' into objfind_qa

debora-pe committed:
joe's comments

debora-pe committed:
small fix

debora-pe commented:
Hi @jhennawi! I addressed your comments and followed your suggestions. Please take a look when you can.
I'll run the tests today or tomorrow. 

I checked that `run_pypeit -s` still shows the objfind plots and it does. Also for `pypeit_coadd_2d`. 

Examples of the QA plots for an echelle and non-echelle spectrographs are above.
The filename of the QA plot includes the basename, the detector number and `slitord_id` (similarly to the spectral flexure correction plots).  E.g., "DE.20170425.50487-dra11_DEIMOS_20170425T140121.014_D07_S0034_obj_prof.png".
Also, if the a negative image exist, a prefix "neg_" or "pos_" is added to the file name. Eg.:
"neg_cN20170331S0216-pisco_GNIRS_20170331T085412.181_D01_S0003_obj_prof.png"
"pos_cN20170331S0216-pisco_GNIRS_20170331T085412.181_D01_S0003_obj_prof.png"

jhennawi reviewed (approved):
This looks great @deborape. Thanks for the very nice work.

debora-pe committed:
Merge branch 'develop' into objfind_qa

debora-pe commented:
Tests pass.
Dev Suite: 
```
--- PYPEIT DEVELOPMENT SUITE PASSED 136/136 TESTS (Masters ignored) ---
Testing Started at 2022-01-18T19:04:26.030363
Testing Completed at 2022-01-19T03:27:04.798528
Total Time: 8:22:38.768165
```

And unit tests
```
pytest . -W ignore
=============================== test session starts ===============================
platform darwin -- Python 3.8.5, pytest-6.1.2, py-1.9.0, pluggy-0.13.1
rootdir: /Users/dpelliccia/PypeIt/PypeIt, configfile: setup.cfg
collected 307 items                                                               

test_alignments.py .                                                        [  0%]
test_arc.py .                                                               [  0%]
test_archive.py ..                                                          [  1%]
test_arcimage.py ..                                                         [  1%]
test_biasframe.py ..                                                        [  2%]
test_bitmask.py ....                                                        [  3%]
test_bpmimage.py ..                                                         [  4%]
test_bspline.py ...............                                             [  9%]
test_buildimage.py ..                                                       [ 10%]
test_calibrations.py .......                                                [ 12%]
test_chk_calibs.py ..                                                       [ 13%]
test_coadd.py .                                                             [ 13%]
test_collate_1d.py ............                                             [ 17%]
test_cooked.py .                                                            [ 17%]
test_coremeta.py .                                                          [ 17%]
test_datacontainer.py ........                                              [ 20%]
test_detector.py ...                                                        [ 21%]
test_display.py .                                                           [ 21%]
test_edgetrace.py .                                                         [ 22%]
test_fitting.py ....                                                        [ 23%]
test_flatfield.py .                                                         [ 23%]
test_flexure.py ..                                                          [ 24%]
test_flux.py ....                                                           [ 25%]
test_fluxspec.py .....                                                      [ 27%]
test_frametype.py ..                                                        [ 28%]
test_history.py ....                                                        [ 29%]
test_load.py .                                                              [ 29%]
test_load_images.py ................                                        [ 34%]
test_manual.py ..                                                           [ 35%]
test_maskdesign_matching.py ...                                             [ 36%]
test_masterframe.py .                                                       [ 36%]
test_match.py ....                                                          [ 38%]
test_metadata.py ......                                                     [ 40%]
test_moment.py ......                                                       [ 42%]
test_msgs.py ..                                                             [ 42%]
test_opticalmodel.py ....                                                   [ 43%]
test_parse.py ...                                                           [ 44%]
test_pca.py ........                                                        [ 47%]
test_processimage.py ..                                                     [ 48%]
test_procimg.py .....                                                       [ 49%]
test_pydl.py .                                                              [ 50%]
test_pypeit.py ..                                                           [ 50%]
test_pypeitimage.py ..                                                      [ 51%]
test_pypeitpar.py ...............................                           [ 61%]
test_pypeitsetup.py ........                                                [ 64%]
test_qa.py .                                                                [ 64%]
test_runpypeit.py ..                                                        [ 65%]
test_scienceimage.py ...                                                    [ 66%]
test_scripts.py ..................                                          [ 71%]
test_setups.py ......................                                       [ 79%]
test_skysub.py ..                                                           [ 79%]
test_slitmask.py ...                                                        [ 80%]
test_slittrace.py ....                                                      [ 82%]
test_spec2dobj.py ......                                                    [ 84%]
test_specobj.py .......                                                     [ 86%]
test_specobjs.py .....                                                      [ 87%]
test_spectrographs.py ....................                                  [ 94%]
test_syncspec.py .                                                          [ 94%]
test_telluric.py ..                                                         [ 95%]
test_traceimage.py .                                                        [ 95%]
test_utils.py ....                                                          [ 97%]
test_vcorr.py ..                                                            [ 97%]
test_wavemodel.py .                                                         [ 98%]
test_wavetilts.py ....                                                      [ 99%]
test_wvcalib.py ..                                                          [100%]

======================== 307 passed in 1208.33s (0:20:08) =========================
```

debora-pe commented:
merging
